scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -175,6 +176,196 @@ def _highly_variable_genes_seurat_v3(
+        in all batches
+    """
+
+    view_to_actual(adata)
+    X = _get_obs_rep(adata, layer=layer)
+    computed_on = layer if layer else 'adata.X'
+
+    # Check for raw counts
+    if check_nonnegative_integers(X) is False:
+        raise ValueError(
+            "`pp.highly_variable_genes` with `flavor='pearson_residuals'`"
+            "expects raw count data."
+        )
+
+    if batch_key is None:
+        batch_info = pd.Categorical(np.zeros(adata.shape[0], dtype=int))
+    else:
+        batch_info = adata.obs[batch_key].values
+    n_batches = len(np.unique(batch_info))
+
+    # Get pearson residuals for each batch separately
+    residual_gene_vars = []
+    for batch in np.unique(batch_info):
+
+        adata_subset = adata[batch_info == batch]
+
+        # Filter out zero genes
+        with settings.verbosity.override(Verbosity.error):
+            nonzero_genes = filter_genes(adata_subset, min_cells=1, inplace=False)[0]
+        adata_subset = adata_subset[:, nonzero_genes]
+
+        if layer is not None:
+            X_batch = adata_subset.layers[layer]
+        else:
+            X_batch = adata_subset.X
+
+        # Prepare clipping
+        if clip == 'auto':
+            n = X_batch.shape[0]
+            clip = np.sqrt(n)
+        if clip < 0:
+            raise ValueError("Pearson residuals require `clip>=0` or `clip='auto'`.")
+
+        if sp_sparse.issparse(X_batch):
+            sums_genes = np.sum(X_batch, axis=0)
+            sums_cells = np.sum(X_batch, axis=1)
+            sum_total = np.sum(sums_genes).squeeze()
+        else:
+            sums_genes = np.sum(X_batch, axis=0, keepdims=True)
+            sums_cells = np.sum(X_batch, axis=1, keepdims=True)
+            sum_total = np.sum(sums_genes)
+
+        # Compute pearson residuals in chunks
+        residual_gene_var = np.ones((X_batch.shape[1])) * np.nan
+        for start in np.arange(0, X_batch.shape[1], chunksize):
+            stop = start + chunksize
+            mu = np.array(sums_cells @ sums_genes[:, start:stop] / sum_total)
+            X_dense = X_batch[:, start:stop].toarray()
+            residuals = (X_dense - mu) / np.sqrt(mu + mu ** 2 / theta)
+            residuals = np.clip(residuals, a_min=-clip, a_max=clip)
+            residual_gene_var[start:stop] = np.var(residuals, axis=0)
+
+        # Add 0 values for genes that were filtered out
+        zero_gene_var = np.zeros(np.sum(~nonzero_genes))
+        residual_gene_var = np.concatenate((residual_gene_var, zero_gene_var))
+        # Order as before filtering
+        idxs = np.concatenate((np.where(nonzero_genes)[0], np.where(~nonzero_genes)[0]))
+        residual_gene_var = residual_gene_var[np.argsort(idxs)]
+        residual_gene_vars.append(residual_gene_var.reshape(1, -1))
+
+    residual_gene_vars = np.concatenate(residual_gene_vars, axis=0)
+
+    # Get cutoffs and define hvgs per batch
+    residual_gene_vars_sorted = np.sort(residual_gene_vars, axis=1)[:, ::-1]
```

Comment:
maybe missing something, but is it needed to sort and resort twice?

Reply:
Thanks for catching this. I think the sort is needed, but the reversal is not!

https://github.com/theislab/scanpy/blob/ff822905ad88be11dee82d760c1c8fdd88066b7f/scanpy/preprocessing/_highly_variable_genes.py#L289-L292

I now removed the reversal, and then use `-n_top_genes` in L291 and `np.greater_equal()` in L292.
Is that what you meant?